packages/hardhat-core/src/internal/core/providers/http.ts

```diff
@@ -76,9 +76,14 @@ export class HttpProvider extends EventEmitter implements EIP1193Provider {
     // the async call that follows would probably loose of the stack trace
     const error = new ProviderError("HttpProviderError", -1);
 
-    const requests = batch.map((r) =>
-      this._getJsonRpcRequest(r.method, r.params)
-    );
+    // we need this to sort the responses
+    const idToIndexMap: Record<string, number> = {};
```

Comment:
I think I'm being too cautious here. We can be quite confident that the ids of the requests are going to be in order (because we call `this._getJsonRpcRequest` serially), so maybe it's enough to just sort the responses by id.

That being said, this approach might be more future proof.